### PR TITLE
If @DisplayName annotation is attached, output the value to test result XML.

### LINKED ISSUE #11445: Junit5 @DisplayName is ignored in junit.xml file
junit.xml file generated by gradle for junit5 tests ignores '@DisplayName' annotation for tests and uses test method name instead
### Expected Behavior
I expect that @DisplayName value is used for '<testcase name=' attribute in generated xml file (when provided, of course)
### Current Behavior
Test method name is used for '<testcase name=' attribute in generated xml file even when @DisplayName exists
### Context
I am going to use CI (Jenkins) in a chain Junit tests -> junit.xml -> CI -> Jira test results. In order to achieve this, I need test case names in junit.xml to be exactly the same as they are stored in Jira (we use zephyr plugin to manage tests): e.g. 'Test Case 1', not 'testCase1()'.

### Steps to Reproduce 
Test class:
```java
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

@DisplayName("Displayed Name for Class")
public class TestClass {
    @Test
    @DisplayName("Displayed Name For Test")
    void test1() {

    }
}
```
build.gradle
```groovy
plugins {
    id 'java'
}

group 'reports'
version '1.0-SNAPSHOT'

sourceCompatibility = 1.8

repositories {
    mavenCentral()
}

test {
    useJUnitPlatform()
}
dependencies {
    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.3.1'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.3.1'
    testImplementation 'org.junit.jupiter:junit-jupiter-params:5.1.0'
}
```
Run gradle test
In build/test-results/test/TEST-TestClass.xml: 
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="TestClass" tests="1" skipped="0" failures="0" errors="0" timestamp="2019-11-22T13:49:44" hostname="localhost" time="0.005">
  <properties/>
  <testcase name="test1()" classname="TestClass" time="0.005"/>
  <system-out><![CDATA[]]></system-out>
  <system-err><![CDATA[]]></system-err>
</testsuite>
```
Expected to have:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="Displayed Name for Class" tests="1" skipped="0" failures="0" errors="0" timestamp="2019-11-22T13:49:44" hostname="localhost" time="0.005">
  <properties/>
  <testcase name="Displayed Name For Test" classname="Displayed Name for Class" time="0.005"/>
  <system-out><![CDATA[]]></system-out>
  <system-err><![CDATA[]]></system-err>
</testsuite>
```
### Your Environment
gradle 6.0.1
Java 13
Junit 5.3.1
Build scan URL:

### PR DESCRIPTION
**Pull request by dhirabayashi**
Fixes #11445

### Context

DisplayName annotation is literally the display name of the test, users would expect the value of DisplayName annotation to be displayed.

https://github.com/gradle/gradle/issues/11445

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by melix** [comment]
Thanks for this PR. Currently I'm unsure we can do this, I need to understand a bit better what are the implications. In particular, if you run `./gradlew testingJvm:intTest` you will notice a number of failing tests.

They are due to the fact that even if there's no `@DisplayName`, the actual name which is used when calling `getClassDisplayName` is _not_ the fully qualifiied class name like `getClassName` but  short name. As a consequence, the file contains:

```xml
<testcase name="inside()" classname="StaticInnerTest$Nested$Nested2" time="0.006"/>
```

instead of:

```xml
<testcase name="inside()" classname="org.gradle.StaticInnerTest$Nested$Nested2" time="0.006"/>
```

The problem is that this file is used by third party tools (Jenkins, ...) and they may rely on this field containing the FQN.

**Comment by dhirabayashi** [comment]
I'm sorry I didn't notice the test failure.

 I guess I should call `getClassDisplayName()` and `getDisplayName())` if `@DisplayName` is given, otherwise call `getClassName()` and `getName()` as before. I'll do some more research to see if that's possible.

**Comment by dhirabayashi** [comment]
Fixed to set the FQN of the class if there is no @DisplayName annotation. Moreover, for the classname attribute of the testcase element, I set the FQN of the class with or without the @DisplayName annotation. Because it is a "classname", I think that it is undesirable to set anything other than the class name.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestDescriptor.java [hunk 1]
```diff
@@ -43,6 +43,12 @@
     @Incubating
     String getDisplayName();
 
+    /**
+     * Whether or not the class has a @DisplayName annotation.
+     * @return true if the class has @DisplayName annotation; false otherwise.
+     */
+    boolean hasDisplayNameAnnotation();
```
Comment (melix) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestDescriptor.java:50]: :x: This shouldn't be part of the public API: the `TestDescriptor` is ecosystem agnostic and it doesn't make sense to talk about a display name annotation here. If we need this, it should be on an internal interface.

Comment (dhirabayashi) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestDescriptor.java]: I've remove the method from the TestDescriptor. I found it to be an unnecessary definition.

**Comment by melix** [comment]
Thanks I'll take another look at it!

**Comment by melix** [comment]
@dhirabayashi because I couldn't update your PR directly I went ahead and created a new one: #13250

Thanks a lot for your contribution!

**Comment by dhirabayashi** [comment]
@melix  I see. Thank you for your quick and precise review!
This was great experience for me.

**Pull request closed without merging**